connectors/kafka/src/main/java/forklift/controller/KafkaController.java

```diff
@@ -0,0 +1,349 @@
+        if (!topics.contains(topic)) {
+            messageStream.addTopic(topic);
+            topics.add(topic);
+            topicsChanged = true;
+            this.notifyAll();
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Unsubscribe the underlying {@link org.apache.kafka.clients.consumer.KafkaConsumer} from the passed in topic and removes the
+     * topic from the message stream.
+     *
+     * @param topic the topic to remove
+     * @return true if the topic was removed, false if it wasn't present
+     */
+    public synchronized boolean removeTopic(String topic) {
+        boolean removed = topics.remove(topic);
+        if (removed) {
+            messageStream.removeTopic(topic);
+            topicsChanged = true;
+        }
+        return removed;
+    }
+
+    /**
+     * Whether or not this service is running.
+     *
+     * @return true if this service is running, else false
+     */
+    public boolean isRunning() {
+        return running;
+    }
+
+    /**
+     * Acknowledge that processing is beginning on a record.  True is returned if this controller is still managing the
+     * partition the record originated from.  If the partition is no longer owned by this controller, false is returned
+     * and the message should not be processed.
+     *
+     * @param record the record to acknowledge
+     * @return true if the record should be processed, else false
+     * @throws InterruptedException if the thread is interrupted
+     */
+    public boolean acknowledge(ConsumerRecord<?, ?> record) throws InterruptedException {
+        AtomicInteger flowCount = flowControl.get(new TopicPartition(record.topic(), record.partition()));
+        if (flowCount != null) {
+            int counter = flowCount.decrementAndGet();
+            if (counter == 0) {
+                synchronized (this) {
+                    this.notifyAll();
+                }
+            }
+        }
+        log.debug("Acknowledge message with topic {} partition {} offset {}", record.topic(), record.partition(), record.offset());
+        return running && this.acknowlegmentHandler.acknowledgeRecord(record);
+    }
+
+    /**
+     * Starts the controller.  Must be called before any other methods.
+     * <p>
+     * <strong>WARNING: </strong>Avoid starting a service which has already run and been stopped.  No attempt is made
+     * to recover a stopped controller.
+     */
+    public void start() {
+        running = true;
+        executor.submit(this::controlLoop);
+    }
+
+    /**
+     * Stops the controller.
+     *
+     * @param timeout  the time to wait for the service to stop
+     * @param timeUnit the TimeUnit of timeout
+     * @throws InterruptedException if the thread is interrupted
+     */
+    public void stop(long timeout, TimeUnit timeUnit) throws InterruptedException {
+        executor.shutdownNow();
+        kafkaConsumer.wakeup();
```

Comment:
Maybe try to `wakeup` the consumer before you try to interrupt it's thread? I would imagine that that would tend to deal with stopping more elegantly.